widget/list.go

```diff
@@ -0,0 +1,413 @@
+package widget
+
+import (
+	"image/color"
+	"math"
+
+	"fyne.io/fyne"
+	"fyne.io/fyne/canvas"
+	"fyne.io/fyne/driver/desktop"
+	"fyne.io/fyne/internal/widget"
+	"fyne.io/fyne/layout"
+	"fyne.io/fyne/theme"
+)
+
+// List is a widget that caches list items for performance and
+// lays the items out in a vertical direction inside of a scroller
+type List struct {
+	BaseWidget
+
+	Length         func() int
+	CreateItem     func() fyne.CanvasObject
+	UpdateItem     func(index int, item fyne.CanvasObject)
+	OnItemSelected func(index int, item fyne.CanvasObject)
+
+	background      color.Color
+	offsetY         int
+	previousOffsetY int
+	selectedItem    *listItem
+	selectedIndex   int
+	itemMin         fyne.Size
+}
+
+// NewList creates and returns a list widget for displaying items in
+// a vertical layout with scrolling and caching for performance
+func NewList(length func() int, createItem func() fyne.CanvasObject, updateItem func(index int, item fyne.CanvasObject)) *List {
+	return &List{BaseWidget: BaseWidget{}, Length: length, CreateItem: createItem, UpdateItem: updateItem, selectedIndex: -1}
+}
+
+// CreateRenderer is a private method to Fyne which links this widget to its renderer
+func (l *List) CreateRenderer() fyne.WidgetRenderer {
+	l.ExtendBaseWidget(l)
+
+	if l.itemMin.Width == 0 && l.itemMin.Height == 0 && l.CreateItem != nil {
+		l.itemMin = newListItem(l.CreateItem(), nil).MinSize()
+	}
+	layout := fyne.NewContainerWithLayout(newListLayout(l))
+	layout.Resize(layout.MinSize())
+
+	scroller := NewVScrollContainer(layout)
+
+	scroller.onOffsetChanged = func() {
+		if l.offsetY == scroller.Offset.Y {
+			return
+		}
+		l.offsetY = scroller.Offset.Y
+		l.BaseWidget.Refresh()
+	}
+
+	objects := []fyne.CanvasObject{scroller}
+	return &listRenderer{BaseRenderer: widget.NewBaseRenderer(objects), list: l, scroller: scroller, layout: layout}
+}
+
+// MinSize returns the size that this widget should not shrink below
+func (l *List) MinSize() fyne.Size {
+	l.ExtendBaseWidget(l)
+
+	return l.BaseWidget.MinSize()
+}
+
+// Refresh updates list to match the current theme
+func (l *List) Refresh() {
+	if l.background != nil {
+		l.background = theme.BackgroundColor()
+	}
```

Comment:
Not sure about this, perhaps `if l.background == nil {`? But is this needed since listRenderer.BackgroundColor() already does this check?

Reply:
Honestly I look at other related widgets to see what they do and this was code that Refresh() in widget/box.go had